### ISSUE TITLE
Json Inference with Dynamically created Pydantic model

### ISSUE THREAD

**Issue by remichu-ai** [Dan-wanna-M/formatron#17]
Hi, i am in the process of integrating your library with my library: https://github.com/remichu-ai/gallama

I am encounter some issue with json generation, where i dynamically create the pydantic schema on the fly depend on user requirement  for function calling and then call it. Currently, lmfe work with the below. Is there any way i can achieve the same with formatron?

This feature will enable me to implement function calling support (which we dont know the pydantic schema ahead)

Script to reproduce my issue is as follow:

```python
from exllamav2 import ExLlamaV2, ExLlamaV2Config, ExLlamaV2Cache, ExLlamaV2Tokenizer
from exllamav2.generator import ExLlamaV2DynamicGenerator, ExLlamaV2Sampler
from formatron.formatter import FormatterBuilder
from formatron.integrations.exllamav2 import create_formatter_filter
from formatron.schemas.pydantic import ClassSchema
from typing import Optional, List, Union, Literal
from typing import Dict, Type

model_dir = "/home/remichu/gallama/models/Mistral-7B-instruct-v0.3-4.5bpw-exl2"
config = ExLlamaV2Config(model_dir)
model = ExLlamaV2(config)
cache = ExLlamaV2Cache(model, max_seq_len=4096, lazy=True)
model.load_autosplit(cache, progress=True)
tokenizer = ExLlamaV2Tokenizer(config)

generator = ExLlamaV2DynamicGenerator(
    model=model,
    cache=cache,
    tokenizer=tokenizer,
)

class Address(ClassSchema):
    street: str
    city: str
    state: Optional[str] = None
    postal_code: str
    country: str
    

class Weather(ClassSchema):
    temperature: int


class ToolCalling(ClassSchema):
    """ The format to call one or multiple tools """
    functions_calling: List[Union[Address, Weather]] = []


def f_get_filter():
    f = FormatterBuilder()
    f.append_line(f"{f.json(ToolCalling, capture_name='json')}")
    exllama_filter = create_formatter_filter(
        generator.model, generator.tokenizer, f)
    return exllama_filter

output = generator.generate(
    prompt="What is Wall Street's address?",
    max_new_tokens=200,
    add_bos=True,
    filters=[f_get_filter()],
)
output

```

The output of above script is as follow:

```
hat is Wall Street\'s address?{ "functions_calling": [ { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "temperature": 0 }, { "t
```

**Comment by Dan-wanna-M** [comment]
@remichu-ai This template works. Can you use this?
```python
def f_get_filter():
    f = FormatterBuilder()
    f.append_line(f"```json\n{f.json(ToolCalling, capture_name='json')}\n```")
    exllama_filter = create_formatter_filter(
        generator.model, generator.tokenizer, f)
    return exllama_filter
```
~~~txt
What is Wall Street's address?```json
{
  "functions_calling": [
    {
      "street": "3 World Financial Center",
      "city": "New York",
      "state": "NY",
      "postal_code": "10281-8098"
    ,
      "country": "United States"
    }
  ]
}
```
~~~

**Comment by remichu-ai** [comment]
It works!

So the format enforcement work and it is just that the model need some help with prompt e.g. ```json to answer better?

Or using ```json have benefit other than just prompting?

**Comment by remichu-ai** [comment]
Another question i have is: I realize formatron doesnt force eos token. Is there an alternative way that i can tell if a generation already ended? I was referring to below code:

```python
def execute():
    prompt = f"""{system_prompt}{inputs[context.index]}{tail}\n\nHere is the extracted information in JSON format:\n\n```\n"""
    output = generator.generate(
        prompt=prompt,
        max_new_tokens=max_new_tokens,
        gen_settings=settings,
        decode_special_tokens=True,
        add_bos=False,
        filters=context.filters,
        completion_only=True,
    )
    # print(repr(output)) # for debug purpose
    context.index += 1
    if context.filters:
        if isinstance(context.filters[0], formatron.integrations.exllamav2.FormatterFilter):
            context.tokens += len(context.filters[0]._formatter._token_ids)
        elif isinstance(context.filters[0], ExLlamaV2TokenEnforcerFilter):
            context.tokens += len(list(filter(lambda x: x != generator.tokenizer.eos_token_id,
                                              context.filters[0].token_sequence)))+1
            # Remove all eos_token_id and supposes only one is generated
            # Currently lm format enforcer forces eos_token_id and does not stop until max_new_token,
            # which does not really contribute to effective tps.
        else:
            raise ValueError(f"Unsupported filter {type(context.filters[0])}")
    else:
        assert not output.endswith(
            generator.tokenizer.eos_token), "Something is wrong"
        context.tokens += max_new_tokens
```

**Comment by remichu-ai** [comment]
Another issue i encounter is, there seems to be some pydantic validation happening inside the filters?

I have the following error which come straight from the ` async for result in job`  it self
```
            job = ExLlamaV2DynamicJobAsync(
                generator=self.pipeline.generator,
                input_ids=input_ids,
                max_new_tokens=max_tokens_to_use,
                gen_settings=settings,
                stop_conditions=stop_conditions,  # self.eos_token_id if self.eos_token_id else None,
                banned_strings=banned_strings,
                decode_special_tokens=True,
                filters=filters,
                token_healing=True,
                identifier=job_id,
            )

            async for result in job:
```

The error looks like this, where seems like ' was used instead of " for celsius.
The challenge here is, the filter should either enforce correct token e.g. ", or just give me the string without raising error so that i can handle the error downstream in my code.
Currently the error happenes at the `async for` which make it a bit hard to handle cause it is inside the exllama generation loop

```
INFO     | mistral:8001 | Full response: {                                                                                                                                                                                                                            | server.py
            "functions_calling": [{"location":"Boston, MA","unit":'celsius'}, {"location":"Hanoi, Vietnam","unit":'celsius'}]                                                                                                                                           
           }                                                                                                                                                                                                                                                            
ERROR    | mistral:8001 | 1 validation error for ToolCalling                                                                                                                                                                                                          | server.py
             Invalid JSON: expected value at line 2 column 56 [type=json_invalid, input_value='{\n "functions_calling":..."unit":\'celsius\'}]\n}', input_type=str]                                                                                                     
               For further information visit https://errors.pydantic.dev/2.8/v/json_invalid  
```

**Comment by Dan-wanna-M** [comment]
> So the format enforcement work and it is just that the model need some help with prompt e.g. ```json to answer better?
> 
> Or using ```json have benefit other than just prompting?

@remichu-ai Yes, the model usually requires a bit more prompting to generate a json, which probably will make the generated json better quality as well. ```json is one of the most simple prompting to make it work. I think it need to be added as a format enforcement rather than in the prompt string though due to token healing issues.



> Another question i have is: I realize formatron doesnt force eos token. Is there an alternative way that i can tell if a generation already ended?

Oh, the idea is that lm format enforcer continuously generates eos_token(so more than one eos_token), which is very inefficient.  Formatron still generates an (exactly one) eos_token(so you can still check eos_token to see if it is finished; `v0.4.2` also supports `is_completed()` method directly).

> Another issue i encounter is, there seems to be some pydantic validation happening inside the filters?

Yes, but this is an internal details and I do not expect this to throw an exception because pydantic validation is usually very lax and Formatron enforces a stricter subset. I will investigate into this bug.

**Comment by Dan-wanna-M** [comment]
It should be fixed in `v0.4.3` now.

**Comment by remichu-ai** [comment]
Awesome. It is working now. 

Let me test further but so far it is fast!
